src/swap/db.rs

```diff
@@ -9,3 +16,757 @@ pub trait Load<T>: Send + Sync + 'static {
+                    .context("Could not write in the DB")?
+                    .context("Stored swap somehow changed, aborting saving")
+            }
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Load<herc20::Funded> for Database {
+    async fn load(&self, swap_id: SwapId) -> anyhow::Result<Option<herc20::Funded>> {
+        let swap = self.get(&swap_id)?;
+
+        Ok(swap.herc20_funded.map(Into::into))
+    }
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+struct Herc20Redeemed {
+    pub transaction: EthereumTransaction,
+    pub secret: Secret,
+}
+
+impl From<Herc20Redeemed> for herc20::Redeemed {
+    fn from(event: Herc20Redeemed) -> Self {
+        herc20::Redeemed {
+            transaction: event.transaction.into(),
+            secret: event.secret,
+        }
+    }
+}
+
+impl From<herc20::Redeemed> for Herc20Redeemed {
+    fn from(event: herc20::Redeemed) -> Self {
+        Herc20Redeemed {
+            transaction: event.transaction.into(),
+            secret: event.secret,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Save<herc20::Redeemed> for Database {
+    async fn save(&self, event: herc20::Redeemed, swap_id: SwapId) -> anyhow::Result<()> {
+        let stored_swap = self.get(&swap_id)?;
+
+        match stored_swap.herc20_redeemed {
+            Some(_) => Err(anyhow!("Herc20 Redeemed event is already stored")),
+            None => {
+                let mut swap = stored_swap.clone();
+                swap.herc20_redeemed = Some(event.into());
+
+                let old_value = serde_json::to_vec(&stored_swap)
+                    .context("Could not serialize old swap value")?;
+                let new_value =
+                    serde_json::to_vec(&swap).context("Could not serialize new swap value")?;
+
+                self.db
+                    .compare_and_swap(swap_id.as_bytes(), Some(old_value), Some(new_value))
+                    .context("Could not write in the DB")?
+                    .context("Stored swap somehow changed, aborting saving")
+            }
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Load<herc20::Redeemed> for Database {
+    async fn load(&self, swap_id: SwapId) -> anyhow::Result<Option<herc20::Redeemed>> {
+        let swap = self.get(&swap_id)?;
+
+        Ok(swap.herc20_redeemed.map(Into::into))
+    }
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+struct Herc20Refunded {
+    pub transaction: EthereumTransaction,
+}
+
+impl From<Herc20Refunded> for herc20::Refunded {
+    fn from(event: Herc20Refunded) -> Self {
+        herc20::Refunded {
+            transaction: event.transaction.into(),
+        }
+    }
+}
+
+impl From<herc20::Refunded> for Herc20Refunded {
+    fn from(event: herc20::Refunded) -> Self {
+        Herc20Refunded {
+            transaction: event.transaction.into(),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Save<herc20::Refunded> for Database {
```

Comment:
:thinking: Maybe we could hide the repetition of all the `Save` implementations behind a macro.

Reply:
Let's clean up during cool down :]